Initial KEP for ambient capabilities in kubernetes.

vinayakankugoyal opened the pull request:
Hello! thank you for taking the time to review his KEP. A lot of the sections are not completed yet but I plan to complete them as we get more consensus on the approach.

vinayakankugoyal commented:
/cc @tallclair

tallclair commented on keps/sig-security/NNNN-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,889 @@
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP proposes that kubernetes provide a way to set ambient capabilities for containers through the Pod manifest. It also proposes changes that must be made to containerd to enable ambient capabilities end-to-end.
```
I assume there's a separate process for proposing features for containerd. It probably makes senes to come to an understanding of what we want in k8s first, but I'd like to see the containerd (and ideally cri-o) changes merge before this KEP is marked implementable.

mikedanese replied:
Are there runtimes other than containerd we should care about? How do we usually expand CRI?

vinayakankugoyal replied:
Currently this KEP focusses on containerd. I think we should expand this to other runtimes in the future. Coordinating this across multiple runtimes at once would be extremely challenging.

vinayakankugoyal replied:
@tallclair 
> but I'd like to see the containerd (and ideally cri-o) changes merge before this KEP is marked implementable

but wouldn't we have to first update the CRI API for them to start reading the AddAmbientCapabilities field?
I think that the order of operation should be:
1. Update CRI API
2. Update containerd and CRI-O
3. Update k8s core APIs

WDYT?

vinayakankugoyal replied:
Yup we have volunteers for reviewing/approving the KEP from 
CRI API - @SergeyKanzhelev 
containerd - @mikebrow
cri-O - @mrunalp

tallclair commented on keps/sig-security/NNNN-ambient-capabilities/README.md:
````diff
@@ -0,0 +1,889 @@
+...
+```
+
+Adding a capability to a non-root container does not add it to the effective and permitted set because effective and permitted sets get cleared when you transition from UID 0 to UID !0. To get around this today users have to apply the capabilities to the binary in their image build phase. 
````
Worth clarifying that this also applies to the default capability set?

vinayakankugoyal replied:
Updated that paragraph PTAL.

tallclair commented on keps/sig-security/NNNN-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,889 @@
+	Capabilities *Capabilities `json:"capabilities,omitempty" protobuf:"bytes,1,opt,name=capabilities"`
+     // The ambient capabilities to add when running containers as non-root.
+     // +optional
+     AmbientCapabilities []Capability `json:"ambientCapabilities,omitempty" protobuf:"bytes,1,rep,name=ambientCapabilities,casttype=[]Capability"`
```
If we go with a separate field, I suggest putting it in the `Capabilities` struct. E.g. `Ambient` works like add, but also makes it ambient.

vinayakankugoyal replied:
Good call! I agree we should do this as well.

tallclair commented on keps/sig-security/NNNN-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,889 @@
+
+### Changes to kubernetes API (https://pkg.go.dev/k8s.io/api/core/v1)
+
+Reusing the existing capabilities field in the securityContext might cause confusion and we propose adding a new field in the SecurityContext called ambientCapabilities. 
```
I would be open to reusing the add field, as long as the _default set_ is still not included in the ambient set.

mikedanese replied:
I think I agree. We inherited the API shape from docker here, and it's not ideal. If we are working in the confines of today's shape, it seems plausible that add: CAP_* on a non-root pod actually means the spec author wanted the process to have CAP_*.

vinayakankugoyal replied:
IIUC what you are suggesting is that add will add the explicitly added capability to the ambient set and the default ones would still be added like before? What I envisioned was that if you are adding a capability using addAmbient then under the hood we drop all the default capabilities and only add the explicitly added capability to the ambient set.

tallclair replied:
Yeah. IIRC, the way `Add` works today is it adds it to the `bounding` & `effective` sets, the same for the default capabilities. What I'm suggesting is that the default capabilities are still only added to bounding & effective (as is proposed), and `Add` adds to those sets, but also to `Ambient`. Does that make sense?

vinayakankugoyal replied:
Default and explicitly added capabilities are added to inherited, permitted, bounding and effective sets. See [this](https://github.com/containerd/containerd/blob/055c801ededcb7a5e82f47bdeed555cdf6c64bd8/oci/spec_opts.go#L818:6) for the code location where this is done. This happens for both default and explicitly added capabilities.

**What in think you are suggesting:** 
* when someone adds a capability explicitly - that capability should be added to all sets i.e. inherited, permitted, bounding, effective and ambient. 
* the default capabilities should only be added to inherited. permitted,bounding and effective sets.
* The advantage of your approach is that:
  * we would not require change to kubernetes API, we would still require changes to the CRI API so that we can distinguish which ones are added to the ambient set vs which ones are not.
* The disadvantage of your approach is:
  * we are changing the behavior of an existing field. (not saying this is bad, I think this shouldn't break anyone)

vinayakankugoyal replied:
I think https://github.com/kubernetes/kubernetes/issues/56374#issuecomment-395814330 talks about why we would want to add a new field.

qiutongs replied:
+1 for adding new field and the only reason to me is that we don't want to break existing customers.

tallclair commented on keps/sig-security/NNNN-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,889 @@
+
+1. AmbientCapabilities can only be added.
+2. If the field is not supplied kubernetes will not add any capabilities to the ambient capabilities set by default.
+3. AmbientCapabilities field should only be set if the Pod is running as non-root. There are 2 ways to handle this:
```
Why? Maybe the container starts as root, but then drops privileges after initialization?

vinayakankugoyal replied:
Good catch! I did not think of this scenario, removed this constraint.

tallclair commented on keps/sig-security/NNNN-ambient-capabilities/README.md:
````diff
@@ -0,0 +1,889 @@
+```
+
+
+Notes about how ambientCapabilities would work:
````
I wonder if we should forbid adding certain capabilities if the container sets `runAsNonRoot`? For example, `runAsNonRoot` is probably not adding much protection if you add `CAP_SYS_ADMIN` (or even `CAP_DAC_OVERRIDE`, in the default set) to the ambient set.

vinayakankugoyal replied:
Yeah I was thinking the same! Great call out!

tallclair commented on keps/sig-security/NNNN-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,889 @@
+	DropCapabilities     []string `protobuf:"bytes,2,rep,name=drop_capabilities,json=dropCapabilities,proto3" json:"drop_capabilities,omitempty"`
+	XXX_NoUnkeyedLiteral struct{} `json:"-"`
+	XXX_sizecache        int32    `json:"-"`
+          // List of ambient capabilities to add.
```
Should `AddAmbientCapabilities` also add them to the inherited & bound sets? I think so?

vinayakankugoyal replied:
I think we would need them to be added to inherited and permitted set.  See https://lwn.net/Articles/636533/. 
TLDR;
AmbientCapabilities obey the invariant that no bit can ever be set in ambient set if it is
not set in both permitted and inheritable.

tallclair commented on keps/sig-security/NNNN-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,889 @@
+	XXX_NoUnkeyedLiteral struct{} `json:"-"`
+	XXX_sizecache        int32    `json:"-"`
+          // List of ambient capabilities to add.
+          AddAmbientCapabilities []string
```
Note: Even if we end up reusing the existing `Add` capabilities field in the k8s API, I agree that it should be kept separate at the CRI level.

tallclair reviewed (commented):
Thanks for picking this up! I'd really like to see this feature.

mrunalp commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,886 @@
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP proposes that kubernetes provide a way to set ambient capabilities for containers through the Pod manifest. It also proposes changes that must be made to containerd to enable ambient capabilities end-to-end.
```
Can you also include CRI-O here? The CRI-O maintainers will make the changes for it. We typically include both containerd and CRI-O for such KEPs.

vinayakankugoyal replied:
I'd be happy to add them if they can confirm that this is possible, during my research I only looked at the containerd code. Is there someone from containerd we can add to review this PR?

vinayakankugoyal replied:
I have updated the scope to include CRI-O as well.

mrunalp commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,886 @@
+	Drop []Capability `json:"drop,omitempty" protobuf:"bytes,2,rep,name=drop,casttype=Capability"`
+  // Ambient capabilities to add
+	// +optional
+	Ambient []Capability `json:"ambient,omitempty" protobuf:"bytes,2,rep,name=ambient,casttype=Capability"`
```
An alternative could be a boolean for ambient that uses the default capabilties list.

vinayakankugoyal replied:
bool does not work because kubernetes adds some capabilities by default and they will all become ambient for a non-root container, essentially making it root. 
it could work though if we don't add default capabilities for non-root containers. (which I think we should do at some point because they don't work anyways 🙃 )

haircommander commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,902 @@
+
+### Changes to kubernetes API (https://pkg.go.dev/k8s.io/api/core/v1)
+
+<<[UNRESOLVED should we reuse add instead of adding new field]>>
```
based on this note https://github.com/kubernetes/kubernetes/issues/56374#issuecomment-395814330
I don't think we should reuse add. I think having an additional slice in the Capabilities struct that is add-only is a clean addition

qiutongs replied:
Here is my two cents about the k8s API design. (Talked with Vinayak offline already)

In one end of the spectrum, k8s API can expose the exact interfaces of the layer below it. In this case, it is [Linux capabilities API](https://man7.org/linux/man-pages/man7/capabilities.7.html). It means k8s API can just allow user directly interact with permitted, inherited, effective, bounding and ambient sets. This way gives user 100% control and flexibility as what Linux allows. This also means k8s API need to adapt with Linux capabilities API. For example, a new thing in Linux capabilities results in a new thing in k8s API.

In the other end, k8s API defines and abstracts its own way of interacting with capabilities. So k8s API can be simpler, at the cost of flexibility. And it should hide details of the underlying OS. The original design of `securityContext.capabilities.add` and `drop` are pretty concise. But it is unclear if it supports both root and non-root. It seems to support but from [this k8s doc](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-capabilities-for-a-container), it doesn't support non root.

> Linux Capabilities: Give a process some privileges, but not all the privileges of the root user.

qiutongs replied:
At this moment, we need to be backward-compatible, which is the only reason that I agree we need a new field. With the new field, I propose to **clarify** the spec, based on [kubernetes/kubernetes#56374 (comment)].

- `add`: if root user, add capabilities; if non-root, behavior is unspecified
- `drop`: if root user, drop capabilities; if non-root, behavior is unspecified
- `setForNonRoot` (open to suggestion, I name it based on *runAsNonRoot*): if non-root user, set capabilities; if root, no ops

My goals are: 
1. clarify the root vs non-root cases 
2. clarify how the new field work with existing ones 
3. hide linux details

qiutongs replied:
Some New thoughts after talked with Vinayak today.

Option 1: Reuse `add` and `drop`
- The capabilities will be added to ambient set, besides the current capability sets. Thus it will work for non-root user case without file capability. 
- It will not break existing scenarios.

Option 2: Add new field - `ambient`
- Users need to understand all those Linux capability concept: https://lwn.net/Articles/636533/
- We need to enhance the `add` and `drop` spec by saying permitted, inheritable, bounding and effective set.

Option 3: Add new field - `addForNonRoot`
- `add` is for any process(parent or child) with root user
- `drop` is for any process(parent or child) with root user
- `addForNonRoot` is for any process(parent or child) with non-root user

dchen1107 commented:
cc/ @SergeyKanzhelev @qiutongs

vinayakankugoyal commented:
/cc @SergeyKanzhelev @qiutongs

ehashman commented:
/cc @mikebrow

ehashman commented:
/ok-to-test

dims commented:
@vinayakankugoyal can you please mention somewhere that "ambient" a word picked from https://man7.org/linux/man-pages/man7/capabilities.7.html thanks!

PushkarJ commented on keps/sig-security/2763-ambient-capabilities/README.md:
````diff
@@ -0,0 +1,920 @@
+      allowPrivilegeEscalation: false
+...
+```
+
````
The KEP might be richer if include the output of error message, for failure scenarios

vinayakankugoyal replied:
Done! Added more deets.

PushkarJ commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,920 @@
+2. Changes described in the [Changes to containerd/containerd (<a href="https://github.com/containerd/containerd">https://github.com/containerd/containerd</a>)](#changes-to-containerdcontainerd-httpsgithubcomcontainerdcontainerd) are made next.
+3. Once a version of `containerd` that supports ambient capabilities is available we make the changes described in the [Changes to kubernetes API (<a href="https://pkg.go.dev/k8s.io/api/core/v1">https://pkg.go.dev/k8s.io/api/core/v1</a>)](#changes-to-kubernetes-api-httpspkggodevk8sioapicorev1) section.
+
+
```
In a real world scenario, different teams are responsible for node OS ( by corollary, containerd or crio) and Kubernetes cluster upgrade and timelines. So in that case, what is the expected behavior when:

1. containerd or crio version that includes this feature is running but Kuberenetes version running, does not include this feature
2. kubernetes version running includes this feature but containerd or crio version running, does not include this feature

vinayakankugoyal replied:
For this feature to work both containerd/crio and kubernetes would need to support the feature. Here's what should happen for both cases:
1. since kuberenetes does not include the feature you cannot create a Pod with the security context that has the ambient capabilities field. 
2. since containerd does not have this feature it should ignore the unknown field in the cri api object. Here is where we would update the documentation to call out that if you are enabling this feature on kubelet you would need containerd/crio version >= x.

sftim replied:
If I try to run a Pod that defines ambient capabilities and that Pod gets scheduled to a node that can't support ambient capabilities:
1. does the whole Pod fail or just the specific container that wanted ambient capabilities?
2. does an Event get created to explain that failure (if so, what event?)
3. does the kubelet keep trying to restart the failed Pod / container?

sftim replied:
Should we define a taint / toleration / [well known label](https://kubernetes.io/docs/reference/labels-annotations-taints/) to mark nodes that can support ambient capability requests?

vinayakankugoyal replied:
> If I try to run a Pod that defines ambient capabilities and that Pod gets scheduled to a node that can't support ambient capabilities:

> 1. does the whole Pod fail or just the specific container that wanted ambient capabilities?
> 2. does an Event get created to explain that failure (if so, what event?)
> 3. does the kubelet keep trying to restart the failed Pod / container?

Since, capabilities are defined per container the container would fail. kubelet would keep retrying the failed container. As this is in Alpha and is opt-in I think that is fine since we would document the supported runtime versions with the feature-flag. But when this feature is on-by default we would probably need to surface why the error hit in kubelet.

> Should we define a taint / toleration / well known label to mark nodes that can support ambient capability requests?

Excellent idea we should definitely do it in beta!

PushkarJ commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,920 @@
+and make progress.
+-->
+- Runtimes other than `containerd` and `CRI-O` are not in scope at the moment.
+
```
For completeness, might be worth calling out that windows workloads are out of scope

vinayakankugoyal replied:
Done!

enj commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,1066 @@
+  Drop []Capability `json:"drop,omitempty" protobuf:"bytes,2,rep,name=drop,casttype=Capability"`
++ // Ambient capabilities to add
++ // +optional
++ Ambient []Capability `json:"ambient,omitempty" protobuf:"bytes,2,rep,name=ambient,casttype=Capability"`
```
Field number is important here.

```suggestion
+ Ambient []Capability `json:"ambient,omitempty" protobuf:"bytes,3,rep,name=ambient,casttype=Capability"`
```

vinayakankugoyal replied:
Done! Good catch.

enj commented on keps/sig-security/2763-ambient-capabilities/README.md:
````diff
@@ -0,0 +1,1066 @@
+
+```diff
+type Capabilities struct {
+  // Added capabilities
+  // +optional
+  Add []Capability `json:"add,omitempty" protobuf:"bytes,1,rep,name=add,casttype=Capability"`
````
Does `Add` serve a meaningful purpose in the API once `Ambient` exists?  Could it be deprecated once `Ambient` is GA?

vinayakankugoyal replied:
good point, I think we could once Ambient is in GA.

invidian replied:
Maybe `Add` make sense when you actually want to retain old behavior and rely on capabilities added during the image building phase?

qiutongs replied:
I am thinking we can dedicate `Add` to root user case, as it currently does.

enj commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,1066 @@
+  XXX_NoUnkeyedLiteral struct{} `json:"-"`
+  XXX_sizecache        int32    `json:"-"`
++ // List of ambient capabilities to add.
++ AddAmbientCapabilities []string `protobuf:"bytes,1,rep,name=add_ambient_capabilities,json=addAmbientCapabilities,proto3" json:"add_ambient_capabilities,omitempty"`
```
```suggestion
+ AddAmbientCapabilities []string `protobuf:"bytes,3,rep,name=add_ambient_capabilities,json=addAmbientCapabilities,proto3" json:"add_ambient_capabilities,omitempty"`
```

vinayakankugoyal replied:
Done! Good catch.

enj commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,1066 @@
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: EnableAmbientCapabilities
+  - Components depending on the feature gate: `kubelet`
```
The API server needs to know about this feature gate as well right - to disallow that `Ambient` field from being used when the feature is disabled?

vinayakankugoyal replied:
IIUC, kube-apiserver should return an error if the ambient field is present in a Pod but the feature-gate is not enabled?

vinayakankugoyal replied:
I updated the list of components to include kube-apiserver as well.

ritazh commented on keps/sig-security/2763-ambient-capabilities/README.md:
````diff
@@ -0,0 +1,1066 @@
+)
+```
+
+We would need to update code at that link to add the ambient capabilities by calling the [WithAmbientCapabilities](https://github.com/containerd/containerd/blob/ab963e1cc16a845567a0e3e971775c29c701fcf8/oci/spec_opts.go#L858) function instead of the [WithoutAmbientCapabilities](https://github.com/containerd/containerd/blob/04f73e3f8a097d95111f8419fa136d196b3a8725/pkg/cri/opts/spec_linux.go#L354) f ...
````
Do we have an issue in containerd to track this proposal?

FYI @cpuguy83

vinayakankugoyal replied:
Yikes, I knew I was forgetting something! Thanks for calling this out. Ill go open one now.

vinayakankugoyal replied:
https://github.com/containerd/containerd/issues/5644

qiutongs replied:
I volunteer to do the work on containerd side. :)

haircommander commented:
LGTM (cri-o maintainer). I like how the changes in this KEP are scoped but also make a clear distinction between ambient vs bounding/effective. feel free to ping me for help getting the cri-o side of this going :+1:

vinayakankugoyal commented on keps/sig-security/2763-ambient-capabilities/README.md:
````diff
@@ -0,0 +1,1066 @@
+2021/06/05 18:21:31 http.ListenAndServe(:80, nil) failed with err: listen tcp :80: bind: permission denied
+```
+
+Capabilities that are either added explicitly in the manifest or by default by kubernetes to a non-root container do not get added to its effective and permitted set because effective and permitted sets get cleared when you transition from UID 0 to UID !0. To get around this today users have to apply the capabilities to the binary in their image build phase.
````
Yeah.

tallclair commented:
Thinking a bit outside the box here: how much demand would there be for this feature if it weren't for NET_BIND_SERVICE? In other words, suppose we just added NET_BIND_SERVICE to the ambient set by default - would there still be demand for adding ambient capabilities support to k8s?

Restricting NET_BIND_SERVICE doesn't make much sense in a kubernetes networking model where each pod has its own IP address. Maybe there are some corner cases with pods mixing trusted & untrusted containers, or for hostNetworking, but I think in most cases having NET_BIND_SERVICE by default would be preferable.

/cc @tabbysable

vinayakankugoyal commented:
I agree that NET_BIND_SERVICE would cover most use cases but...
Why not provide the means to run containers as non-root while only adding the capabilities that are needed. I think there are a few system monitoring agents that could benefit from it (but I guess you could argue that they should be running as root?)
The point is that the work to make NET_BIND_SERVICE ambient by default is probably the same as either what this KEP proposes if we decide to do the defaulting in k8s (i.e. via kubelet and updating the CRI API) or is equivalent to https://github.com/containerd/containerd/pull/5668 where we add every explicitly added capability to the ambient set (instead of every we would add just NET_BIND_SERVICE).
Giving support for ambient capabilities would allow more workloads to run as non root but one could argue that are they really running as non-root give the capabilities they could have?

tallclair commented:
> Why not provide the means to run containers as non-root while only adding the capabilities that are needed.

- Additional complexity (surfaced to the user) - Changing & maintaining API surface isn't cheap. If there isn't strong demand / use cases, we should avoid it.
- Confusion / policy complication - There are already users who assume that as long as your container isn't `privileged`, it is safe. To really check if a container is "contained" you also need to look at host namespaces, host paths, capabilities, etc. I'm worried ambient capabilities could do the same for non-root (it's running as non-root, so it must be safe). You already need to set `AllowPriviligeEscalation: false` (I wish we could have rolled that into `runAsNonRoot` instead).

> Giving support for ambient capabilities would allow more workloads to run as non root but one could argue that are they really running as non-root give the capabilities they could have?

Yeah, for a lot of these capabilities I think the benefit of running as non-root is pretty marginal if you're adding a powerful capability. In a lot of ways, it might be better to just keep running as root and drop the capabilities you don't need.

If we started with just setting the default, we could always revisit exposing the knobs to the user in the future.

vinayakankugoyal commented:
* Additional complexity surfaced to the user - this could be avoided by reusing the existing add and drop API.
* policy complication - I 100% agree with you on this one. Running as non-root with CAP_SYS_ADMIN is equal to running as root (although the KEP does propose not allowing such capabilities). For alpha I think just adding NET_BIND_SERVICE to the ambient set might be a fair compromise. But instead of adding it by default I would like to propose adding it to the ambient set only when it is explicitly added (that way we are not giving capabilities that aren't needed). We can make this change completely in the container runtime(s). WDYT?

Jc2k commented:
From my point of view, I don't want to control ambient capabilities as much as I want `add` to do what you'd expect it to do from it's name. So I agree with @vinayakankugoyal that reusing the existing add/drop would be best.

I think the current situation is confusing. To a simple naive end user like myself it feels like a bug. That confusion is at least "fail safe" - my code is running with less capabilities than i expected rather than extra ones. But it still confusing and frustrating.

Personally if I have some code that I want to run with `NET_ADMIN` and only `NET_ADMIN`, I would rather have kubernetes set up a sandbox with those restrictions in place rather than trust the application to do it. As a cluster administrator I am more inclined to trust the kubernetes sandbox than random code from another part of the hypothetical org.

Being able to audit capabilities with manifest checking tools is also something that you could loose if you go down the route of making us use a root helper script to set caps and drop uid.

So my take would be to use the existing add/drop fields rather than adding new API. Let the user decide what to do with it, gently encourage the use of `PodSecurityPolicy`. Or maybe this feature requires you to turn it on with a `PodSecurityPolicy`?

Random-Liu commented:
> Thinking a bit outside the box here: how much demand would there be for this feature if it weren't for NET_BIND_SERVICE? In other words, suppose we just added NET_BIND_SERVICE to the ambient set by default - would there still be demand for adding ambient capabilities support to k8s?

Agree with @tallclair here. If most cases we see are just for `NET_BIND_SERVICE`, we could consider including it into the default ambient CAP set for non-root container, instead of making ambient set configurable in the Kubernetes API.

I think features and API changes should be driven by use cases. We may only want to add ambient cap in the Kubernetes API when there is a use case that a capability is needed for some non-root containers, but doesn't make sense to grant it to all non-root containers.

It does make sense to grant `NET_BIND_SERVICE` to all non-root containers to me (if we are OK with the behavior change and rollout it properly). However, if there is another use case that requires configurable ambient cap, we can revisit exposing the API.

invidian commented:
>So my take would be to use the existing add/drop fields rather than adding new API. Let the user decide what to do with it, gently encourage the use of PodSecurityPolicy. Or maybe this feature requires you to turn it on with a PodSecurityPolicy?

Note, that PSPs are now deprecated and IIUC there is no official alternative for it yet.

>I think features and API changes should be driven by use cases. We may only want to add ambient cap in the Kubernetes API when there is a use case that a capability is needed for some non-root containers, but doesn't make sense to grant it to all non-root containers.

While I agree on API changes driven by use cases, do note that introducing new API which is too limited will be confusing for the users and may require quick further changes. I'd at least make sure whatever we agree on is possibly extendable.

Jc2k commented:
I think I have seen more need for NET_ADMIN than for BIND_SERVICE personally.

But if it's really not a bug that you can't use "add" with a non root container, can there be an error message if you try to do it? Rather than if looking like it's done something but hasn't, which is what happened when I tried a while ago. This is one of the reasons why this feels like a bug report than a feature request.

Random-Liu commented:
> While I agree on API changes driven by use cases, do note that introducing new API which is too limited will be confusing for the users and may require quick further changes.

I think what was proposed is to not expose the API for now, and the default ambient cap set for non-root user can be defined in Kubernetes.

> We can make this change completely in the container runtime(s).

If possible, I hope we can still add the CRI field, and contain the behavior in Kubernetes. I assume we may need to define the default set of ambient cap list, and disallowed caps etc.

vinayakankugoyal commented:
Here is what I propose:
1. We reuse the k8s API i.e. keep only add and drop in the container SecurityContext.Capabilities
2. We update the CRI API to make it explicit which capabilities we want to be ambient.
3. When a user adds a set of allowed capabilities to a **non-root** container, we explicitly add those to the CRI API ambient field. 
4. We document what this set of allowed capabilities is that a user can add to a non-root container. (it could be only NET_BIND_SERVICE and NET_ADMIN for now)

I don't want to add capabilities by default to a non-root container, when the user doesn't explicitly add it to the securityContext.

Random-Liu commented:
I'm OK with 2 and 3. Especially for 2, I think making it explicit would give the container runtime the chance to report error if ambient cap is not supported, e.g. by the host kernel or container runtime.

For 1, I don't have a strong opinion. Personally, I think `Add` is reasonable to me if we decide to expose the API.

For 4, can we have a more concrete use case about `NET_ADMIN` before adding it?

tallclair commented:
I strongly agree with all @Random-Liu's points. I agree that whichever way we go with this, starting with an explicitly defined CRI API sounds like a good idea.

For @vinayakankugoyal's proposal in https://github.com/kubernetes/enhancements/pull/2757#issuecomment-870940352, I'm worried about (3). I was in favor of reusing the `add` field for ambient capabilities under the assumption that they would all be added, but if only a subset of them get added to the ambient set I'm afraid that would get pretty confusing.

Responding to a few other comments:

> Note, that PSPs are now deprecated and IIUC there is no official alternative for it yet.

We have an alternative in the works: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/2579-psp-replacement/README.md

> I think I have seen more need for NET_ADMIN than for BIND_SERVICE personally.

You've seen that for non-root containers? That's surprising to me, since I think of that as a pretty elevated permission. I'd be very interested in seeing some examples, if you have any references.

> But if it's really not a bug that you can't use "add" with a non root container, can there be an error message if you try to do it? 

Under the current implementation, it's still useful as long as you're using file capabilities (and not setting `allowPrivilegeEscalation=false`).

qiutongs commented:
I think the discussion around NET_BIND_SERVICE is about whether we want to have allowed/disallowed cap set in the **long term**. If not, I strongly suggest let's not have it from the beginning. 

I propose to **NOT** have special care to ANY capability. **Whatever customer provides in the `add`, we add to ambient set.** If customer use CAP_SYS_ADMIN, it is their misuse. (We can warn them).  This seems to align with Tim's idea in https://github.com/kubernetes/enhancements/pull/2757#issuecomment-870994018 ?

This will make the `securityContext.capabilities.add` spec works as it should have.

Random-Liu commented:
> I was in favor of reusing the add field for ambient capabilities under the assumption that they would all be added, but if only a subset of them get added to the ambient set I'm afraid that would get pretty confusing.

I also feel like we should not ignore other caps.

> I propose to NOT have special care to ANY capability. Whatever customer provides in the add, we add to ambient set. If customer use CAP_SYS_ADMIN, it is their misuse. (We can warn them). This seems to align with Tim's idea in #2757 (comment) ?

It doesn't necessarily mean that we have to support ALL caps, we can just add an API validation for the `Capabilities` fields for non-root containers, and reject pods with disallowed capabilities specified. Loosening the restriction is always easier than hardening it in the future.

tallclair commented:
> we can just add an API validation for the Capabilities fields for non-root containers, and reject pods with disallowed capabilities specified

Not if we're reusing the existing add field, since this would be tightening validation.

Random-Liu commented:
> Not if we're reusing the existing add field, since this would be tightening validation.

I thought we are treating the pod yaml with non-root + cap before as a misusage, and that is why we can change behavior for this kind of pod?
* 1st way of changing behavior is to blindly apply those caps now, which I'm not sure whether that would cause unexpected result, e.g. a non-root pod with very privileged CAPs that users added by mistake;
* 2nd way of changing behavior is to fail explicitly;
* 3rd way of changing behavior is to just apply a selected set of caps, which is confusing to users.

If we don't want any of these, we will need a separate field.

tabbysable commented:
> > But if it's really not a bug that you can't use "add" with a non root container, can there be an error message if you try to do it?
> 
> Under the current implementation, it's still useful as long as you're using file capabilities (and not setting `allowPrivilegeEscalation=false`).

This my favorite piece of k8s trivia: adding caps on a non-root container is useful even with `allowPrivilegeEscalation=false` as long as the entrypoint binary in the container has the file capabilities you're adding. :ghost:

Much of this discussion talks about the behavior of `add` being different for root and non-root containers. How would we tell the difference at API validation time? I assume we'd have to look at `mustRunAsNonRoot`, because the apiserver has no idea whether a textual `runAsUser` is root or not. (This is my least-favorite thing about `mustRunAsNonRoot` -- it has to fail at the last minute, which is bad UX.) Because this is so awkward, I don't like the idea of behaving differently for root vs non-root.

I think it's fine for this UI to be tightly coupled to the Linux capabilities feature - non-Linux runtimes are not going to map well enough to make re-using the capabilities fields safe anyway.

Because the UI design is so tightly tied to Linux capabilities anyway, anything other than adding a new field for ambient caps is going to be confusing in some way or another.

If we want to add another field but make it simpler, I guess it could be `Ambient: true|false` rather than another array... but I think the array is best because at least it's straightforward for users who understand Linux capabilities. They're very complex, and I don't think we're in a place to make them simpler without accidentally causing a security issue for somebody.

Random-Liu commented:
> If we want to add another field but make it simpler, I guess it could be Ambient: true|false rather than another array... but I think the array is best because at least it's straightforward for users who understand Linux capabilities. They're very complex, and I don't think we're in a place to make them simpler without accidentally causing a security issue for somebody.

From a low level implementation perspective, `containerd` erase all ambient caps for all containers https://github.com/containerd/containerd/blob/f0a32c66dad1e9de716c9960af806105d691cd78/pkg/cri/server/container_create_linux.go#L238, but today, but add caps for all other caps https://github.com/containerd/containerd/blob/f0a32c66dad1e9de716c9960af806105d691cd78/oci/spec_opts.go#L823-L827.

So the `Add` and `Drop` only takes effects on `Bounding`, `Effective`, `Permitted` and `Inheritable`. And `Ambient` is **always empty**.

In this case, defining a new field just adding extra `Ambient` caps makes the behavior crystal clear. And given the problems with reusing `Add` in https://github.com/kubernetes/enhancements/pull/2757#issuecomment-871047487, I also feel like a new field is better.

vinayakankugoyal commented:
> If we want to add another field but make it simpler, I guess it could be Ambient: true|false rather than another array... but I think the array is best because at least it's straightforward for users who understand Linux capabilities. They're very complex, and I don't think we're in a place to make them simpler without accidentally causing a security issue for somebody.

Agreed! adding a new array field (like originally proposed) LGTM!

> From a low level implementation perspective, containerd erase all ambient caps for all containers https://github.com/containerd/containerd/blob/f0a32c66dad1e9de716c9960af806105d691cd78/pkg/cri/server/container_create_linux.go#L238, but today, but add caps for all other caps https://github.com/containerd/containerd/blob/f0a32c66dad1e9de716c9960af806105d691cd78/oci/spec_opts.go#L823-L827. @mrunalp What is cri-o behavior today?

CRI-O also only adds capabilities to the inhetiable,bounding,effective and permitted set. https://github.com/cri-o/cri-o/blob/741bd7c8c16f495efa94a836a01097eebd21280d/server/container_create.go#L287. Ambient set is always empty. So essentially the behavior is the same.

> And based on my understanding Ambient caps are useful for root containers as well? They can control what caps they give to children processes that are non-privileged. And that cap list may not be the same as the root caps in Add?

Yes this will let us limit what capabilities can be added through the ambient field without worrying about breaking or complicating current scenarios.

sftim commented:
As well as this API change, I'd like to be able to set ambient capability defaults at the namespace level. The default default should be no ambient capabilities.

sftim commented on keps/sig-security/2763-ambient-capabilities/kep.yaml:
```diff
@@ -0,0 +1,50 @@
+# The following PRR answers are required at alpha release
+# List the feature gate name and the components for which it must be enabled
+feature-gates:
+  - name: EnableAmbientCapabilities
```
Usually we don't put `Enable` in [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) names - otherwise they would all collate under “E”.

vinayakankugoyal replied:
Updated to AmbientCapabilities.

sftim commented on keps/sig-security/2763-ambient-capabilities/README.md:
````diff
@@ -0,0 +1,1075 @@
+      command: ["simpleserver", "--port", "8080"]
+```
+
+```bash
````
nit: This isn't bash (it's commands mixed with sample output)

vinayakankugoyal replied:
Updated.

sftim commented on keps/sig-security/2763-ambient-capabilities/README.md:
````diff
@@ -0,0 +1,1075 @@
+      command: ["simpleserver", "--port", "80"]
+```
+
+```bash
````
nit: This isn't bash (it's commands mixed with sample output)

vinayakankugoyal replied:
Updated.

sftim reviewed (commented):
Some feedback - a mix of formatting nits and other thoughts

vinayakankugoyal commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,1075 @@
+
+1. AmbientCapabilities can only be added.
+2. Default capabilities are not added to the ambient capabilities set.
+3. Only the capability that is explicitly added in the manifest will be added to the ambient set.
```
Thanks updated.

vinayakankugoyal commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,1075 @@
+Author proposes that we go with `Option 1` above as it keeps the API simple, without requiring to understand what sets their capability needs to be added to. If a user adds a capability in the manifest it just works.
+<<[/UNRESOLVED]>>
+
+Notes about how ambientCapabilities would work:
```
Updated.

vinayakankugoyal commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,1075 @@
+-->
+- Runtimes other than `containerd` and `CRI-O` are not in scope at the moment.
+- Windows is not covered in this design.
+
```
Good suggestion. Updated!

vinayakankugoyal commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,1075 @@
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Enable support for ambient capbilities for kubernetes and the `containerd` and `CRI-O` runtime.
```
Updated.

qiutongs commented:
>  but I think the array is best because at least it's straightforward for users who understand Linux capabilities. They're very complex, and I don't think we're in a place to make them simpler without accidentally causing a security issue for somebody.

I really like this principle and hope it can guide the design of k8s API. That being said, I agree we have an additional `Ambient` field. We should also explicitly call out the current `Add`  adds capability to "permitted, effective, bounding, inherited" sets.  I cannot accept part of the capability is explicit Linux but the other part is vague. This is a reinforcement of current k8s API specification, not just improving comments.

qiutongs commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,1076 @@
+
+Some capabilities like `CAP_SYS_ADMIN` and `CAP_DAC_OVERRIDE` make a non-root user very powerful and we should restrict them from being added to the ambient capabilities set.
+
+This is out of scope for alpha release.
```
Based on what Lantao said, we should be restrictive at the beginning. So maybe we want this?

Random-Liu replied:
Yeah, we can just add some api validations based on my understanding.

The alpha should be more restrictive to me. Opening more caps could be out of scope for alpha release instead.

qiutongs commented:
/lgtm

vinayakankugoyal commented:
@sftim 

> If I try to run a Pod that defines ambient capabilities and that Pod gets scheduled to a node that can't support ambient capabilities:

> does the whole Pod fail or just the specific container that wanted ambient capabilities?
The container would fail.
> does an Event get created to explain that failure (if so, what event?)
The behavior would be same as running a container as rootless and not having the required capabilities added. The container logs would contain an error like `standard_init_linux.go:211: exec user process caused "operation not permitted"`
> does the kubelet keep trying to restart the failed Pod / container?
Yes it does.
```

vinayakankugoyal commented:
@sftim 
> Should we define a taint / toleration / well known label to mark nodes that can support ambient capability requests?

Great idea! I think a node label like kubernetes.io/node-supports-ambient-capabilities (probably need a better name) would be great.

ehashman commented:
/assign

I can take PRR for this one when it's ready, although from https://github.com/kubernetes/enhancements/issues/2763#issuecomment-914484690 it looks like it might slip?

ehashman commented:
/retitle KEP 2763: Ambient capabilities in Kubernetes

vinayakankugoyal commented:
@tallclair and @tabbysable - per the discussion in the SIG Security meeting I have updated this KEP. Does it look good to be merged as provisional?

tabbysable commented:
Per discussion at SIG Security meeting, approving this in provisional state because the UNRESOLVED sections seem to capture the necessary next steps to move it beyond provisional.

tabbysable commented:
/approve

tabbysable commented:
/label tide/merge-method-squash

vinayakankugoyal commented:
@tabbysable could you please LGTM again (sorry for the inconvenience). There was a typo in the kep metadata that was causing the verify-kep-metadata.sh test to fail. Those are the only changes since last LGTM. Thank you!

tabbysable commented:
we should probably bump the milestones to 1.24, since this is certainly not going into 1.23, just to avoid confusion

tabbysable commented:
alright everyone, buckle up:
/lgtm

vinayakankugoyal commented:
/retest

giuseppe commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,929 @@
+There are 2 options here:-
+- **Option 1:** Reuse Add field in [Capabilities](https://pkg.go.dev/k8s.io/api/core/v1#Capabilities)
+
+  When a capability gets added explicitly to a non-root container it also gets added to the ambient set in addition to getting added to inheritable, permitted, bounding and effective sets. The default capabilities are not added to the ambient set.
```
having capabilities being added to the inheritable set turned out to be a CVE affecting multiple runtimes (Moby, containerd, CRI-O, Podman...), more details in the Moby advisory: https://github.com/moby/moby/security/advisories/GHSA-2mm7-x5h6-5pvq

Having the capabilities added directly to the ambient set requires them to be present in the inheritable set as well, thus getting back the old behavior.

I think this option should not be considered.

Jc2k commented on keps/sig-security/2763-ambient-capabilities/README.md:
```diff
@@ -0,0 +1,929 @@
+
+#### Restricted ambient capabilities.
+
+<<[UNRESOLVED what is the set of capabilities that we should allow to be made ambient]>>
```
Previously when this was discussed PSP's were deprecated and there wasn't alternative. But we now have a PodSecurity admission controller in beta and enabled by default as of 1.23.

So I think if there are any restrictions it should be done in the PodSecurity admission controller. Most should probably be under the 'Privileged' category.

sftim replied:
@Jc2k are you thinking of a change to the [pod security standards](https://kubernetes.io/docs/concepts/security/pod-security-standards/)? These are what you can enforce with Pod Security admission.

Jc2k replied:
It depends.

For context, let me stress again that when this section of the KEP were first discussed, the admission controller wasn't in beta (if it existed at all) so it wasn't considered a valid solution to this `UNRESOLVED` part of the KEP.

In the case where we re-use `securityContext.capabilities.add`, the existing standards look like they are valid for root and non-root containers. Neither would be allowed to use `CAP_SYS_ADMIN` or `CAP_DAC_OVERRIDE` etc unless they were using the `privileged` policy level. This is good.

If we are **not** reusing `securityContext.capabilities.add`, then I think the standards should be updated anyway. They currently cover all the ways you can add capabilities to containers. If there is a new way to add a capabitility to a container, it should be covered by the standards.

Jc2k commented:
I'm aware of at least one popular CNI that now has a non-root mode. It does this through suid binaries. I'd much prefer a DaemonSet with an explicit and limited capability allow list to a container image using suid binaries, even if both ultimately have powerful permission sets.
